beacon_node/network/src/network_beacon_processor/rpc_methods.rs

```diff
@@ -479,4 +603,210 @@ impl<T: BeaconChainTypes> NetworkBeaconProcessor<T> {
+                );
+                return;
+            }
+        };
+
+        let oldest_blob_slot = self
+            .chain
+            .store
+            .get_blob_info()
+            .oldest_blob_slot
+            .unwrap_or(data_availability_boundary_slot);
+        if request_start_slot < oldest_blob_slot {
+            debug!(
+                self.log,
+                "Range request start slot is older than data availability boundary.";
+                "requested_slot" => request_start_slot,
+                "oldest_blob_slot" => oldest_blob_slot,
+                "data_availability_boundary" => data_availability_boundary_slot
+            );
+
+            return if data_availability_boundary_slot < oldest_blob_slot {
+                self.send_error_response(
+                    peer_id,
+                    RPCResponseErrorCode::ResourceUnavailable,
+                    "blobs pruned within boundary".into(),
+                    request_id,
+                )
+            } else {
+                self.send_error_response(
+                    peer_id,
+                    RPCResponseErrorCode::InvalidRequest,
+                    "Req outside availability period".into(),
+                    request_id,
+                )
+            };
+        }
+
+        let forwards_block_root_iter =
+            match self.chain.forwards_iter_block_roots(request_start_slot) {
+                Ok(iter) => iter,
+                Err(BeaconChainError::HistoricalBlockError(
+                    HistoricalBlockError::BlockOutOfRange {
+                        slot,
+                        oldest_block_slot,
+                    },
+                )) => {
+                    debug!(self.log, "Range request failed during backfill";
+                        "requested_slot" => slot,
+                        "oldest_known_slot" => oldest_block_slot
+                    );
+                    return self.send_error_response(
+                        peer_id,
+                        RPCResponseErrorCode::ResourceUnavailable,
+                        "Backfilling".into(),
+                        request_id,
+                    );
+                }
+                Err(e) => {
+                    self.send_error_response(
+                        peer_id,
+                        RPCResponseErrorCode::ServerError,
+                        "Database error".into(),
+                        request_id,
+                    );
+                    return error!(self.log, "Unable to obtain root iter";
+                        "request" => ?req,
+                        "peer" => %peer_id,
+                        "error" => ?e
+                    );
+                }
+            };
+
+        // Pick out the required blocks, ignoring skip-slots.
+        let mut last_block_root = req.start_slot.checked_sub(1).and_then(|prev_slot| {
+            self.chain
+                .block_root_at_slot(Slot::new(prev_slot), WhenSlotSkipped::Prev)
```

Comment:
So the `ForwardsBlockRootsIterator` is used here and that uses `WhenSlotSkipped::Prev` functionality. That's the reason for creating `maybe_block_roots` manually. Don't think we have an alternative at present but it might be nice to add `WhenSlotSkipped` configuration to `ForwardsBlockRootsIterator`.

I think we need `WhenSlotSkipped::Prev` in this specific line too because `last_block_root` is used to determine whether the first slot of the request is a skip (the first slot in the `ForwardsBlockRootsIterator`). So setting `WhenSlotSkipped::None` and having this return `last_block_root = None` would tell us the prior slot was a skip, but we'd still have no way of knowing whether the first slot of the request is a skip.

Added a comment here: 9960580394a1635d42920f5a45cb56a568e46ccb